Disable iPad Mini UA overrides when Site Specific Hacks are disabled

brentfulgham opened the pull request:
#### d146f360321a12c140eb7278b3d173900d44fcb8
<pre>
Disable iPad Mini UA overrides when Site Specific Hacks are disabled
<a href="https://bugs.webkit.org/show_bug.cgi?id=281212">https://bugs.webkit.org/show_bug.cgi?id=281212</a>
<a href="https://rdar.apple.com/50035167">rdar://50035167</a>

Reviewed by Devin Rousso and Patrick Angle.

iPad Mini UA overrides needs to be disabled when Site Specific Hacks
are disabled.

Karl&apos;s work on that issue revealed that changing the Site Specific Quirks
state through the WebInspector did not trigger a notification in the UIProcess.
Since the UIProcess governs the load, quirks about using Desktop versus
Mobile state were not disabled.

This patch fixes both issues.

It also updates the generated preferences to use an optional override value
when appropriate, similar to how the WebCore&apos;s Settings overrides work. This
will allow the page to snap back to default behavior when the WebInspector
is disconnected.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WebCore/inspector/InspectorClient.h:
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::disable):
(WebCore::InspectorPageAgent::overrideSetting):
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesDefinitions.h.erb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesGetterSetters.cpp.erb:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::setDeveloperPreferenceOverride):
* Source/WebKit/UIProcess/WebPreferences.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::isDesktopClassBrowsingRecommended const):

Canonical link: <a href="https://commits.webkit.org/285166@main">https://commits.webkit.org/285166@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/d9bf4f3ba9e48a860797f0aee185a48d0ceaf6bb

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/71588 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/51001 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/24362 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/75700 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/22793 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/73703 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/58802 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/22613 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/56543 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/15017 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/74654 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/46273 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/61663 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/36992 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/71096 "Passed tests") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/42936 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/19129 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/21134 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/64713 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/64840 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/19493 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/77418 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/70837 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/15822 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/18672 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/64256 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/15865 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/61698 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/64252 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/12399 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/6037 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/92623 "Built successfully") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/11001 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/46801 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/1580 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/20415 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/47872 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/49156 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/47614 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

patrickangle commented on Source/WebCore/inspector/InspectorClient.h:
```diff
@@ -44,6 +44,7 @@ enum class InspectorClientDeveloperPreference : uint8_t {
     PrivateClickMeasurementDebugModeEnabled,
     ITPDebugModeEnabled,
     MockCaptureDevicesEnabled,
+    NeedsSiteSpecificQuirks,
```
Maybe `SiteSpecificQuirksEnabled` to match the other preferences here?

brentfulgham replied:
The naming `NeedsSiteSpecificQuirks` is used all over the WebInspector code base, so I would prefer to avoid having to change all of the JS and other support files that expect this naming.

dcrousso commented on Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
```diff
@@ -759,6 +759,11 @@ void WebInspectorUIProxy::setDeveloperPreferenceOverride(WebCore::InspectorClien
+
+    case InspectorClient::DeveloperPreference::NeedsSiteSpecificQuirks:
+        if (RefPtr inspectedPage = m_inspectedPage.get())
+            inspectedPage->protectedPreferences()->setNeedsSiteSpecificQuirks(overrideValue && overrideValue.value());
```
rather than modify the preference directly, can we have this be an override like all the others so that when Web Inspector is closed we're able to go back to whatever behavior it had before?

karlcow replied:
@dcrousso would there be a possibility for people on iOS to be able to disable Site Specific Quirks through a menu in addition to Web Inspector, in order to allow people testing sites on mobile but do not have access to a macbook.

brentfulgham replied:
@karlcow That mechanism is controlled outside of WebKit code, so would need to be changed elsewhere.

brentfulgham replied:
> rather than modify the preference directly, can we have this be an override like all the others so that when Web Inspector is closed we're able to go back to whatever behavior it had before?

That's a great idea! I'll do so and get a new patch up.

brentfulgham replied:
Updated patch with this change, @dcrousso .

dcrousso replied:
@karlcow i coulda sworn there was a toggle for it somewhere in iOS/Safari settings, but i can't seem to find it

regardless, the general principle of Web Inspector is that anything changed while debugging should be undone as soon as Web Inspector is disconnected, with the only exceptions being HTML, CSS, and JS (i.e. changes to page content can persist but engine/browser adjustments must be undone)

dcrousso commented on Source/WebKit/UIProcess/WebPageProxy.cpp:
```diff
@@ -15419,6 +15419,17 @@ Ref<BrowsingContextGroup> WebPageProxy::protectedBrowsingContextGroup() const
+
+bool WebPageProxy::needsSiteSpecificQuirks() const
+{
+    return m_needsSiteSpecificQuirksOverride ? *m_needsSiteSpecificQuirksOverride : m_preferences->needsSiteSpecificQuirks();
```
this slightly concerns me because there's no way to ensure that future developers use `WebPageProxy::needsSiteSpecificQuirks` instead of `WebPreferences:: needsSiteSpecificQuirks`

ideally, we'd have a similar system for `WebPreferences` like what already exists for `Settings` where there's an `std::optional<bool> m_fooInspectorOverride;` that's checked inside the getter (see [208371@main](https://commits.webkit.org/208371@main) for more info)

but if this needs to get in ASAP then i dont have any issues with it as-is (tho i would appreciate it if you did this work right after or created a bug so we know to do it soon)

brentfulgham replied:
Oh! I didn't know about that mechanism. Let me see if I can figure out how to hook this in properly.

brentfulgham replied:
@dcrousso Update the patch to use the same mechanism you had for generated WebCore Settings. I'm a little surprised that so many items in the `UnifiedWebPreferences.yaml` file have `inspectorOverride: true`, but don't seem to be part of `InspectorPageAgent::overrideSetting`. Maybe none of these are ever changed in the Inspector and need to be reflected in the UI Process.

Still, I would imagine that things like `CompositingBordersVisible` and `CompositingRepaintCountersVisible` would need to seen in the UIProcess, but I think they tweak `defaults` storage, so perhaps don't need to be explicitly IPC'd to the UIProcess.

It seems like maybe `ICECandidateFilteringEnabled` should be part of `InspectorPageAgent::overrideSetting`, but maybe no UI was created for that? (@youennf: Do you know?)

dcrousso replied:
yeah i think we only bothered sending things to the UIProcess that would actually be used in the UIProcess.  would be neat to generalize it across all settings/preferences Web Inspector can touch (i.e. all of `Inspector::Protocol::Page::Setting`)

we definitely could also use this for `ICECandidateFilteringEnabled` and/or `MockCaptureDevicesEnabled` as well if you have the time to do so :)

brentfulgham replied:
I think for that, I will do a separate patch. :-)

dcrousso reviewed (approved):
nice work!  thanks for iterating with me :)

brentfulgham committed:
Disable iPad Mini UA overrides when Site Specific Hacks are disabled
https://bugs.webkit.org/show_bug.cgi?id=281212
rdar://50035167

Reviewed by Devin Rousso and Patrick Angle.

iPad Mini UA overrides needs to be disabled when Site Specific Hacks
are disabled.

Karl's work on that issue revealed that changing the Site Specific Quirks
state through the WebInspector did not trigger a notification in the UIProcess.
Since the UIProcess governs the load, quirks about using Desktop versus
Mobile state were not disabled.

This patch fixes both issues.

It also updates the generated preferences to use an optional override value
when appropriate, similar to how the WebCore's Settings overrides work. This
will allow the page to snap back to default behavior when the WebInspector
is disconnected.

* Source/WTF/Scripts/GeneratePreferences.rb:
* Source/WebCore/inspector/InspectorClient.h:
* Source/WebCore/inspector/agents/InspectorPageAgent.cpp:
(WebCore::InspectorPageAgent::disable):
(WebCore::InspectorPageAgent::overrideSetting):
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesDefinitions.h.erb:
* Source/WebKit/Scripts/PreferencesTemplates/WebPreferencesGetterSetters.cpp.erb:
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
* Source/WebKit/UIProcess/Inspector/WebInspectorUIProxy.cpp:
(WebKit::WebInspectorUIProxy::setDeveloperPreferenceOverride):
* Source/WebKit/UIProcess/WebPreferences.h:
* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::isDesktopClassBrowsingRecommended const):

Canonical link: https://commits.webkit.org/285166@main
